pkg/kubelet/secret_manager.go

```diff
@@ -39,3 +54,181 @@ func newSimpleSecretManager(kubeClient clientset.Interface) (secretManager, erro
+	if len(items) != 2 {
+		return "", "", fmt.Errorf("invalid key: %s", key)
+	}
+	return items[0], items[1], nil
+}
+
+func (s *secretStore) Add(namespace, name string) {
+	key := keyFunc(namespace, name)
+	secret, err := s.kubeClient.Core().Secrets(namespace).Get(name)
+	if err != nil {
+		glog.Errorf("Unable to retrieve a secret %s/%s: %v", namespace, name, err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.secret = secret
+		item.refCount++
+	} else {
+		s.items[key] = &secretStoreItem{secret: secret, refCount: 1}
+	}
+}
+
+func (s *secretStore) Delete(namespace, name string) {
+	key := keyFunc(namespace, name)
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		item.refCount--
+		if item.refCount == 0 {
+			delete(s.items, key)
+		}
+	}
+}
+
+func (s *secretStore) Get(namespace, name string) (*api.Secret, error) {
+	key := keyFunc(namespace, name)
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		return item.secret, nil
+	}
+	return nil, fmt.Errorf("secret no present")
+}
+
+func (s *secretStore) Refresh() {
+	s.Lock()
+	keys := make([]string, 0, len(s.items))
+	for key := range s.items {
+		keys = append(keys, key)
+	}
+	s.Unlock()
+
+	secrets := make([]*api.Secret, 0, len(keys))
+	for _, key := range keys {
+		namespace, name, err := namespaceNameFunc(key)
+		if err != nil {
+			glog.Errorf("Unexpected key transform error: %v", err)
+			secrets = append(secrets, nil)
+			continue
+		}
+		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name)
+		if err != nil {
+			glog.Errorf("Unable to retrieve a secret %s/%s: %v", namespace, name, err)
+			secret = nil
+		}
+		secrets = append(secrets, secret)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for i, key := range keys {
+		if secrets[i] == nil {
+			// Ignore updating to nil.
+		}
+		if item, ok := s.items[key]; ok {
+			item.secret = secrets[i]
+		}
+	}
+}
+
+// cachingSecretManager keeps a cache of all secrets necessary for registered pods.
+// It implements the following logic:
+// - whenever a pod is created or updated, the current versions of all its secrets
+//   are grabbed from apiserver and stored in local cache
+// - every GetSecret call is served from local cache
+// - every X seconds we are refreshing the local cache by grabbing current version
+//   of all registered secrets from apiserver
+type cachingSecretManager struct {
+	secretStore *secretStore
+
+	sync.Mutex
+	registeredPods map[string]*api.Pod
+}
+
+func newCachingSecretManager(kubeClient clientset.Interface) (secretManager, error) {
+	return &cachingSecretManager{
+		secretStore:    newSecretStore(kubeClient),
+		registeredPods: make(map[string]*api.Pod),
+	}, nil
+}
+
+func (c *cachingSecretManager) GetSecret(namespace, name string) (*api.Secret, error) {
+	return c.secretStore.Get(namespace, name)
+}
+
+func (c *cachingSecretManager) RegisterPod(pod *api.Pod) {
+	for _, reference := range pod.Spec.ImagePullSecrets {
```

Comment:
I may be missing something, but I think the only other thing that kubelet is currently looking into is envVars:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L478
What do you mean by volumes here?

Though, when refreshing secrets, we are only looking into the "real ones":
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L604
Is it a bug?

Is this a bug?